### ISSUE TITLE
Move Production  DNS to Notify AWS Accounts

### ISSUE THREAD

**Issue by ben851** [cds-snc/notification-planning-core#36]
## Description

As a developer/operator of Notify, I want to be able to control the DNS flow for GC Notify in my own infrastructure as code, and not have a dependency on an external team.

#### WHY
Currently the Production DNS is owned by CDS SRE, and we are required to submit PRs to their code base in order to modify these references. In Staging, this is done via click-ops and slack requests. It is important to codify both production and staging DNS entries in order to ensure consistency. At the same time, it is also important to remove dependencies on external teams to create high velocity development and operation cycles.

#### WHAT
We can open a PR with CDS SRE to have the *.notification.<staging or production parent domain> delegated to our own AWS accounts, at which point we can codify our DNS entries in our own terraform repositories. 

#### VALUE
By having full control over our DNS entries, we will streamline the release and change management process. We will also be able to quickly spin up new environments and automatically create DNS entries for them under the sandbox dns zone. 

## Acceptance Criteria

- [x] Both DNS zones are delegated to their respective AWS accounts
- [x] All DNS entries for notify are done using Terraform
- [x] Terraform pipelines in github actions manage the DNS as well

## QA Steps
- [ ] Verify that the website works
- [ ] DNS entries resolve correctly 
- [ ] New environment is spun up with automatic DNS endpoints created

**Comment by ben851** [comment]
This is a pretty quick change, but we will need to be careful w/ Production.

**Comment by ben851** [comment]
Attaching this to the BCP epic as I need to get this done in order to automate the validation of ACM certificates.

**Comment by ben851** [comment]
* New PR up for review

**Comment by ben851** [comment]
* Staging released 
* Will migrate DNS for staging tomorrow morning

**Comment by ben851** [comment]
* Staging DNS Migrated
* Moving to blocked while we wait for SRE to give access to Prod.

**Comment by ben851** [comment]
* Had to implement some magic for the DKIM validation records

**Comment by ben851** [comment]
* I deployed a fix to staging to correct the issue with DKIM validation
* AWS says it may take up to 72 hours to take effect. Will check at the end of the week
*

**Comment by ben851** [comment]
Will follow up with SRE today

**Comment by ben851** [comment]
* I contacted Calvin yesterday with no response. Guillaume Charest is the owner of the ticket, but is on vacation.

**Comment by ben851** [comment]
* Calvin says he needs to talk w/ Max about this and will get back to me.

**Comment by ben851** [comment]
* Calvin has said that the SRE team wants us to move towards requesting a name server redirect with SSC so we can manage it in our own account. Will have to investigate the process for this :(

**Comment by ben851** [comment]
* Alternate option will not work due to DNS technical limitations
* I've sent an email to the SSC account managers for CDS asking for the procedure to get a request in with the SSC DNS team.

**Comment by ben851** [comment]
* SSC Responded saying they will set up a meeting. Have not heard back, will poke today.

**Comment by ben851** [comment]
* Emailed SSC on Friday. Have not heard back. Will ask again.

**Comment by ben851** [comment]
* Emailed SSC Monday, no response.

**Comment by ben851** [comment]
* SSC Confirmed that I should open a ticket with TBS help desk which will in turn be forwarded to SSC. 
* Ticket has been opened.

**Comment by jimleroyer** [comment]
Ticket was closed; we need to open a form with Principal Publisher. Ben is in process to submit the new form.

**Comment by ben851** [comment]
* I've opened a request w/ the principal publisher, and informed SSC liasons.

**Comment by jimleroyer** [comment]
Need to re-raise the ticket as this was closed while Ben was on vacation. We know of next steps though and will contact proper contacts.

**Comment by jimleroyer** [comment]
To do a Terraform release this morning with Ben to move DNS setup to production.

**Comment by jimleroyer** [comment]
We encountered issues while running the plan to production. The new DNS provider wasn't behaving as expected and we had to make further refactoring to accommodate the differences with our different environments.

**Comment by ben851** [comment]
DNS zone is now deploying to production. Needs to be tested and still need integration w/ SSC.

**Comment by ben851** [comment]
We had a heck of a time getting SSC and ESDC together to deploy this.

Instead, we are looking at providing terraform with access to the notification.canada.ca route53 zone already owned by CDS.

**Comment by ben851** [comment]
Opened an issue with SRE to get access to the notification.canada.ca route 53 zone
https://github.com/cds-snc/dns/issues/395

**Comment by ben851** [comment]
After speaking with Pat some more, I'm going to kill two birds with one stone and move the terraform plan/apply workflows to OIDC authentication.

**Comment by ben851** [comment]
Had to revert the OIDC because it was causing problems with quicksight. Investigating

**Comment by ben851** [comment]
Will be debugging today

**Comment by ben851** [comment]
Refactored OIDC into the new multi-job workflows. Reproduced bug with quicksight, added an additional permission for pull-requests on the github workflow, and it's been resolved. Need 2 PRs approved:

Staging fix:
https://github.com/cds-snc/notification-terraform/pull/1421

Production:
https://github.com/cds-snc/notification-terraform/pull/1419

**Comment by ben851** [comment]
Staging and Production running on OIDC again. 

Sylvia is working on [this ticket](https://github.com/cds-snc/dns/issues/395) today to grant us access to the prod DNS account, at which point I will be doing diffs and imports to migrate our stuff over.

**Comment by jimleroyer** [comment]
Ben will work on a Terraform release as the OIDC prod changes that were done did not work. He will fix this. Afterward, we will be waiting on Sylvia to unblock us.

**Comment by ben851** [comment]
OIDC fixed in prod, had to open an issue with SRE to get increased permissions on the notification-terraform-plan role in prod.

**Comment by ben851** [comment]
PR for new role here:
https://github.com/cds-snc/dns/pull/397

I commented that we also need at least read access for the notify-core team.

**Comment by ben851** [comment]
I got access on Friday - I will work on this today.

**Comment by ben851** [comment]
Did a comparison between "real"prod and the "fake" prod that DNS records are set to. There were a few discrepancies - merged in some missing entires for

doc.notification.canada.ca
document.notification.canada.ca
api.document.notification.canada.ca (maybe this is why doc-download-api didn't work in dev
documetnation.notification.canada.ca
www.notification.canada.ca

Also had a mismatch on the weighted api.notification.canada.ca which in real life points to itself instead of the api-gateway lambda endpoint. That PR will be merged today.

Once both are merged, I will re-compare between real and fake, and if all is good, change the provider in TF to point to the "real" DNS and start doing imports.

**Comment by P0NDER0SA** [comment]
Ben is not here today, we will see where it is tomorrow

**Comment by ben851** [comment]
Finished checks and did an import on prod Friday. I then made a backup of these import states and then restored the old states. Need to merge this PR to staging and then create a prod release, and restore the new states
https://github.com/cds-snc/notification-terraform/pull/1447

**Comment by ben851** [comment]
Merged to staging, production release ready.

https://github.com/cds-snc/notification-terraform/pull/1449

**Comment by P0NDER0SA** [comment]
Just approved this one

**Comment by ben851** [comment]
Implemented in prod. Need to get SRE to remove their references to DNS in their repository

**Comment by ben851** [comment]
Issue opened with SRE
https://github.com/cds-snc/dns/issues/408
